### PR TITLE
[Messenger] Removing "sync" transport and replacing it with config trick

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no (but removes some hackiness)
| New feature?  | no
| Deprecations? | yes (BC break yes)
| Tickets       | Helps #34066
| License       | MIT
| Doc PR        | not needed

I introduced `SyncTransport` in 4.3 as a way to allow people to, for example, route a message to an `async` transport and then override that to a `sync` transport in the dev environment. The implementation was always a bit of a hack, and makes this one transport behave different than others *and* makes messages routed to this transport behave a *bit* different than sync messages that are simply not routed at all.

Thanks for an idea from Nicolas, this fixes that. Gone are `SyncTransport`. Instead, the `sync://` transport string is still supported, but it's just a "config" trick: messages routed to transports that are "sync" are simply filtered out of the "senders" in the DI extension. Nice & clean.

### CODE REVIEW

File: src/Symfony/Component/Messenger/CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,8 @@ CHANGELOG
 4.4.0
 -----
 
+ * [BC BREAK] The `SyncTransport` and `SyncTransportFactory` classes
+   were removed.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/CHANGELOG.md]: no line split would be good here

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 2]
```diff
@@ -1803,30 +1812,25 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
...
-                if (!isset($senderReferences[$sender])) {
+                if (isset($senderReferences[$sender])) {
+                    $realSenders[] = $sender;
+                } elseif (!\in_array($sender, $syncTransports, true)) {
                     throw new LogicException(sprintf('Invalid Messenger routing configuration: the "%s" class is being routed to a sender called "%s". This is not a valid transport or service id.', $message, $sender));
                 }
             }
 
-            $messageToSendersMapping[$message] = $messageConfiguration['senders'];
+            if (0 !== \count($realSenders)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: `if ($realSenders) {`

**Review by nicolas-grekas — COMMENTED** [review]
sweet, less code !

**Review by nicolas-grekas — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @weaverryan.

**Comment by Devristo** [comment]
This seems to break the case when the transport name is defined using an environment variable:

Works:
```
framework:
   messenger:
     transports:
       async: "sync://"
```


Broken when `MESSENGER_TRANSPORT=sync://`:
```
framework:
   messenger:
     transports:
       async: "%env(MESSENGER_TRANSPORT)%"
```

Yields `No transport supports the given Messenger DSN "sync://"`

**Pull request merged**